### PR TITLE
[Serve] Detect http.disconnect can cancel handle requests

### LINKED ISSUE #21425: [Bug] [Serve] HTTPProxy High Memory Usage
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Serve

### What happened + What you expected to happen

```

ray.exceptions.RayTaskError(RayOutOfMemoryError): ray::RayServeWrappedReplica.handle_request() (pid=220, ip=10.16.0.131, repr=<ray.serve.replica.0 object at 0x7f55eab83390>)
ray._private.memory_monitor.RayOutOfMemoryError: More than 90% of the memory on node ses-lb4tlh5uyqcyg4kxkmxneppy-head-node-type-7f586984db-d5hf2 is used (22.36 / 23.52 GB). The top 10 memory consumers are:
PID	MEM	COMMAND
29465	15.39GiB	ray::HTTPProxyActor
220	4.72GiB	ray::0
188	1.27GiB	ray::ServeController.listen_for_change()
37	0.25GiB	bash -l -i health_check.sh
123	0.14GiB	/home/ray/anaconda3/bin/python -u /home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/agen
51	0.11GiB	/home/ray/anaconda3/bin/python -u /home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/dash
8	0.07GiB	/home/ray/anaconda3/bin/python /home/ray/anaconda3/bin/ray start --head --no-monitor --port=6379 --o
98	0.05GiB	/home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet --raylet_socket_name=
36	0.05GiB	/home/ray/anaconda3/bin/python /home/ray/anaconda3/bin/jupyter-notebook --NotebookApp.token=4aa0b60d
In addition, up to 0.0 GiB of shared memory is currently being used by the Ray object store.
---
--- Tip: Use the `ray memory` command to list active objects in the cluster.
--- To disable OOM exceptions, set RAY_DISABLE_MEMORY_MONITOR=1.
---
2022-01-06 08:44:06,427	ERROR import_thread.py:88 -- ImportThread: Connection closed by server. component=serve deployment=ComposedModel replica=ComposedModel#bgQROq
```

### Versions / Dependencies

image

Might be related to ray.wait?

Ray 1.9.0

### Reproduction script

See above

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR should fix the symptom observed from #21425. The hypothesis is as follows:
- Some condition lead to client sending many requests in parallel to Serve.
- The application cannot handle such high load, so requests are all queued. 
- Client timed out and aborted connections, kicking a retry loop.
- Client send even more requests, leading to even more requests to added to the queue. 
- Serve is now working through a queue containing mostly requests that's no longer needed. 

The fix is simply to detect the http.disconnect message and cancel the `handle.remote` task. We used asyncio.Task's [cancel mechanism](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel) will will raise an exception on next `await` within the task. 

## Related issue number
Closes #21425

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Also manually verified that HTTPProxy's memory did increase before the fix and didn't after the fix.

### CODE REVIEW

File: python/ray/serve/http_proxy.py [hunk 1]
```diff
@@ -54,11 +56,30 @@ async def _send_request_to_handle(handle, scope, receive, send):
 
     retries = 0
     backoff_time_s = 0.05
+    loop = asyncio.get_event_loop()
+    # We have received all the http request conent. The next `receive`
+    # call might never arrive; if it does, it can only be `http.disconnect`.
```
Comment (edoakes) [inline python/ray/serve/http_proxy.py:61]: is it possible to add a check for the http disconnect case?

Comment (edoakes) [reply python/ray/serve/http_proxy.py:61]: also, are we sure this is guaranteed?

File: python/ray/serve/http_proxy.py [hunk 2]
```diff
@@ -54,11 +56,30 @@ async def _send_request_to_handle(handle, scope, receive, send):
...
+            assignment_task.cancel()
         try:
+            object_ref = await assignment_task
             result = await object_ref
+            client_disconnection_task.cancel()
             break
+        except asyncio.CancelledError:
+            # Here because the client disconnected, we will return a custom
+            # error code for metric tracking.
+            return DISCONNECT_ERROR_CODE
```
Comment (edoakes) [inline python/ray/serve/http_proxy.py:82]: Where is this used?

**Review by edoakes — APPROVED** [review]
For my own understanding, what are the cases in which we get the HTTP disconnect message?

File: python/ray/serve/http_proxy.py (hunk 1 above)
Comment (simon-mo) [reply python/ray/serve/http_proxy.py:61]: Yup. But I can add a check.

File: python/ray/serve/http_proxy.py (hunk 2 above)
Comment (simon-mo) [reply python/ray/serve/http_proxy.py:82]: it's a metric tag

**Comment by simon-mo** [comment]
> For my own understanding, what are the cases in which we get the HTTP disconnect message?

When a client "hang up" before we send the response to the client. It manifests as tcp socket disconnect/drop.

**Pull request merged**